File: src/storage/dedup.py
"""Hybrid deduplication strategy for usage snapshots."""

import time
from typing import Dict, Optional


class DedupTracker:
    """Tracks last stored snapshots for deduplication."""

    HEARTBEAT_INTERVAL = 60  # Store every 60s even if no change

    def __init__(self):
        """Initialize tracker."""
        self._last_stored = {}  # service -> {metric_name: (used_pct, timestamp)}

    def should_store(
        self,
        service: str,
        metric_name: str,
        used_pct: int
    ) -> bool:
        """
        Determine if snapshot should be stored.

        Stores if:
        1. Value changed from last stored value, OR
        2. >= 60s since last store (heartbeat)

        Args:
            service: Service name ('claude' or 'codex')
            metric_name: Metric name (e.g., 'session', '5h')
            used_pct: Current usage percentage

        Returns:
            True if snapshot should be stored
        """
        key = f"{service}:{metric_name}"
        now = time.time()

        # Check if we have previous data for this metric
        if key not in self._last_stored:
            # First time seeing this metric, store it
            self._last_stored[key] = (used_pct, now)
            return True

        last_pct, last_time = self._last_stored[key]

        # Check if value changed
        if used_pct != last_pct:
            self._last_stored[key] = (used_pct, now)
            return True

        # Check if heartbeat interval elapsed
        if (now - last_time) >= self.HEARTBEAT_INTERVAL:
            self._last_stored[key] = (used_pct, now)
            return True

        # No change and within heartbeat interval, skip
        return False

    def should_store_metrics(
        self,
        service: str,
        metrics: Dict
    ) -> bool:
        """
        Check if any metric in the dict should be stored.

        Args:
            service: Service name
            metrics: Metrics dict

        Returns:
            True if at least one metric should be stored
        """
        for metric_name, metric_data in metrics.items():
            # Skip subscription_type key
            if metric_name == 'subscription_type' or not isinstance(metric_data, dict):
                continue

            used_pct = metric_data.get('used_pct', 0)

            if self.should_store(service, metric_name, used_pct):
                return True

        return False

    def clear(self):
        """Clear tracking data."""
        self._last_stored.clear()

File: packages/core/src/storage/dedup.ts
/**
 * Deduplication tracker to avoid storing duplicate snapshots.
 * Ported from Python src/storage/database.py (DedupTracker class)
 */
import type { MetricsDict } from "../types.js";

export class DedupTracker {
  static readonly HEARTBEAT_INTERVAL = 60; // seconds

  private lastStored = new Map<string, [number, number]>();

  shouldStore(service: string, metricName: string, usedPct: number): boolean {
    const key = `${service}:${metricName}`;
    const now = Date.now() / 1000;

    const entry = this.lastStored.get(key);
    if (!entry) {
      this.lastStored.set(key, [usedPct, now]);
      return true;
    }

    const [lastPct, lastTime] = entry;

    if (usedPct !== lastPct) {
      this.lastStored.set(key, [usedPct, now]);
      return true;
    }

    if (now - lastTime >= DedupTracker.HEARTBEAT_INTERVAL) {
      this.lastStored.set(key, [usedPct, now]);
      return true;
    }

    return false;
  }

  shouldStoreMetrics(service: string, metrics: MetricsDict): boolean {
    for (const [metricName, metricData] of Object.entries(metrics)) {
      if (metricName === "subscription_type" || typeof metricData !== "object" || metricData === null) {
        continue;
      }
      const usedPct = metricData.used_pct ?? 0;
      if (this.shouldStore(service, metricName, usedPct)) {
        return true;
      }
    }
    return false;
  }

  clear(): void {
    this.lastStored.clear();
  }
}
